File: 2018/day07.py
import collections
import fileinput
import heapq

from utils.graphs import toposort


MAX_WORKERS = 5
BASE_STEP_TIME = 60


def p1(pairs):
    edges = collections.defaultdict(set)
    for a, b in pairs:
        edges[a].add(b)

    return toposort(
        edges.keys(),
        lambda x: edges[x],
    )


def p2(pairs):
    edges = collections.defaultdict(set)
    indegrees = collections.defaultdict(int)
    for a, b in pairs:
        edges[a].add(b)
        indegrees[b] += 1
    working_nodes = [node for node in edges.keys() if indegrees[node] == 0]
    heapq.heapify(working_nodes)
    tasks = {}

    t = 0
    while working_nodes or tasks:
        for node, end_time in list(tasks.items()):
            if t >= end_time:
                del tasks[node]
                for edge in edges[node]:
                    indegrees[edge] -= 1
                    if indegrees[edge] == 0:
                        heapq.heappush(working_nodes, edge)

        while working_nodes and len(tasks) < MAX_WORKERS:
            node = heapq.heappop(working_nodes)
            tasks[node] = t + BASE_STEP_TIME + 1 + ord(node) - ord("A")
        t += 1

    return t - 1


def parse_line(line):
    words = line.split()
    return (words[1], words[7])


if __name__ == "__main__":
    lines = [parse_line(line) for line in fileinput.input()]
    print("Part 1: %s" % "".join(p1(lines)))
    print("Part 2: %s" % p2(lines))

File: 2018/cpp/day07.cc
#include <string>
#include <unordered_set>
#include <utility>
#include <vector>

#include "graphs.h"
#include "utils.h"

int MAX_WORKERS = 5;
int BASE_STEP_TIME = 60;

namespace {
using std::pair;
using std::string;
using std::vector;
}  // namespace

string p1(vector<pair<char, char>> pairs) {
  std::unordered_set<char> values;
  std::unordered_map<char, std::vector<char>> edges;
  for (const auto& p : pairs) {
    values.insert(p.first);
    values.insert(p.second);
    edges[p.first].push_back(p.second);
  }

  std::function<vector<char>(const char&)> neighbors = [&edges](const char& v) {
    return edges[v];
  };

  const auto result =
      aoc::graphs::toposort({values.begin(), values.end()}, neighbors).value();

  return string(result.begin(), result.end());
}

int p2(vector<pair<char, char>> pairs) {
  std::unordered_map<char, std::vector<char>> edges;
  std::unordered_map<char, int> indegrees;
  for (const auto& p : pairs) {
    edges[p.first].push_back(p.second);
    ++indegrees[p.second];
  }

  std::priority_queue<char, std::vector<char>, std::greater<char>>
      working_values;
  for (const auto& p : edges) {
    if (indegrees[p.first] == 0) {
      working_values.push(p.first);
    }
  }

  std::unordered_map<char, int> tasks;
  int t = 0;
  while (working_values.size() || tasks.size()) {
    for (auto it = tasks.begin(); it != tasks.end();) {
      char value = it->first;
      int end_time = it->second;
      if (t >= end_time) {
        tasks.erase(it++);
        for (auto edge : edges[value]) {
          --indegrees[edge];
          if (indegrees[edge] == 0) {
            working_values.push(edge);
          }
        }
      } else {
        ++it;
      }
    }
    while (working_values.size() && tasks.size() < MAX_WORKERS) {
      auto value = working_values.top();
      working_values.pop();
      tasks[value] = t + BASE_STEP_TIME + 1 + value - 'A';
    }
    ++t;
  }

  return t - 1;
}

int main() {
  vector<string> lines = aoc::utils::getlines();

  vector<pair<char, char>> pairs;
  std::transform(
      lines.begin(), lines.end(), std::back_inserter(pairs),
      [](const string& line) { return pair<char, char>(line[5], line[36]); });

  std::cout << "Part 1: " << p1(pairs) << std::endl;
  std::cout << "Part 2: " << p2(pairs) << std::endl;
}
